### ISSUE TITLE
Number of atom types not updated when adding interstitial atoms to the lattice

### ISSUE THREAD

**Issue by QLRO** [pierrehirel/atomsk#25]
Dear Pierre,

<p>First thank you for making this fantastic software that helps me a lot in my work. </p>

<p>I would like to report a bug regarding adding atoms to an existing LAMMPS file.</p>

<h1>Step to recreate</h1>

1. atomsk --create bcc 2.856 Fe orient [100] [010] [001] -duplicate 2 2 2 lammps  
2. atomsk Fe.lmp -add-atom H random 111 lammps  
3. Give the new file a name and check the created LAMMPS file

<h1>Potential bug</h1>
LAMMPS data file requires defining number of atom types at the beginning. The file created in the previous step will leave the line "1  atom types" unchanged, where it should be changed to "2  atom types". This will give rise to an error when reading from the file.

<h1>Suggested change</h1>
<p>Change the line whenever new atom type is added when modifying a LAMMPS file.</p>

<p>Merry Christmas and happy new year!</p>

Best regards,
QLRO

**Comment by QLRO** [comment]
Just realized all new atoms are numbered to type 1 as well. So maybe the line change is not sufficient.

**Comment by pierrehirel** [comment]
Dear Sir,

Thank you for pointing out what was, and what not really, and is no longer, a bug in Atomsk. Let me explain.

What you describe is true: is the property "type" exists, then when new atoms are added ("-add-atoms"), they are created with the wrong "type". This can indeed be considered as a bug.

However, this bug only occurs because intermediate files are written in the LAMMPS data format. This format is _not_ intended for intermediate files, it is intended for running a simulation with LAMMPS. This format contains "types", so in your second command Atomsk reads those "types" and pass them to the option "-add-atoms", which does not deal with them correctly.

I always recommend to use a different file format for intermediate files, for example the CFG format which is quite versatile:

```
atomsk --create bcc 2.856 Fe orient [100] [010] [001] -duplicate 2 2 2 Fe.cfg
atomsk Fe.cfg -add-atom H random 111 FeH.lmp
```

Notice how if you do that, the "bug" does not appear at all. The option "-add-atoms" will not mess things up, because the atom "types" are not defined in the CFG file. When writing the final LAMMPS data file "Fe.lmp", Atomsk will create new types, which will be different for each atomic species.

In summary, it's only a bug if you use LAMMPS data format for intermediate files, which you should not do.

Better yet, if you can then avoid using intermediate files, and use a single command:

```
atomsk --create bcc 2.856 Fe orient [100] [010] [001] -duplicate 2 2 2 -add-atom H random 111 FeH.lmp
```

My final point, is that this specific bug was corrected in the latest version on GitHub. In this version the option "-add-atoms" checks if "types" are defined. If new atoms are of the same species as pre-existing atoms then they are given the same type, otherwise a new "type" is created. I invite you to download this version and compile it, and tell me if it solves your problem.

Best regards

**Comment by QLRO** [comment]
Thank you for the reply. I was indeed using the version downloaded from the website. Sorry 
for not checking the Github change before posting.

**Comment by Dreamer-zh** [comment]
Dear Dr. Pierre:
Thank you so much for replying to me.

In fact, the intermediate file in the .lmp format does have a problem. When I changed the intermediate file to .cfg format, I have checked that the atom type and the type in the atom coordinates are correct. I found the original problematic command:
_atomsk --create bcc 2.8553 Fe -duplicate 5 5 2 FeBox.lmp
atomsk FeBox.lmp -add-atom Ti random 1 FeTi.lmp

atomsk FeBox.lmp -select random 10 any Fe -substitute Fe Ti FeTiBox.lmp
atomsk FeBox.lmp -select random 1% Fe -substitute Fe Ti FeTiBox.lmp_
Change the intermediate file format to .cfg and it has been solved

Supplement: The Atomsk software version I use is atomsk_b0.10.6_Linux-amd64.tar.gz

Happy New Year!
Best regards
